File: latihan/Latihan 1/KoleksiKendaraan.cpp
#include "KoleksiKendaraan.hpp"
#include <iostream>

using namespace std;

/* CTOR - Default */
KoleksiKendaraan::KoleksiKendaraan(){
    size = 100;
    nEff = 0;
}

/* CTOR - User Defined */
KoleksiKendaraan::KoleksiKendaraan(int size, int nEff){
    this->size = size;
    this->nEff = nEff;
}

/* CCTOR */
KoleksiKendaraan::KoleksiKendaraan(KoleksiKendaraan &other){
    this->size = other.size;
    this->nEff = other.nEff;
}

/* Copy Asignment */
KoleksiKendaraan KoleksiKendaraan::operator=(KoleksiKendaraan &other){
    return KoleksiKendaraan(other);
}


/* DTOR */
KoleksiKendaraan::~KoleksiKendaraan(){}

void KoleksiKendaraan::printAll(){
    cout << "Ukuran Koleksi: " << size << endl;
    cout << "Jumlah Kendaraan: " << nEff << endl;
    for(int i = 0; i < nEff; i++){
        koleksi[i].printInfo();
    }
}

void KoleksiKendaraan::operator<<(const Kendaraan &kendaraan){
    if(nEff < size){
        nEff++;
        koleksi.push_back(kendaraan);
    }
}


void KoleksiKendaraan::operator<<(const KoleksiKendaraan &koleksiKendaraan){
    for(int i = 0; i < koleksiKendaraan.nEff; i++){
        if(nEff < size){
            nEff++;
            koleksi.push_back(koleksiKendaraan.koleksi[i]);
        } else{
            break;
        }
    }
}

File: latihan/latihan-2/KoleksiKendaraan.java
public class KoleksiKendaraan {
    private int size;
    private int nEff;
    private Kendaraan[] koleksi;

    /* CTOR */
    public KoleksiKendaraan(){
        size = 100;
        nEff = 0;
        koleksi = new Kendaraan[size];
    }

    /* CTOR - User Defined */
    public KoleksiKendaraan(int _size){
        size = _size;
        nEff = 0;
        koleksi = new Kendaraan[_size];
    }

    /* CCTOR */
    public KoleksiKendaraan(KoleksiKendaraan other){
        koleksi = new Kendaraan[other.size];
        size = other.size;
        nEff = other.nEff;
        for(int i = 0; i < other.nEff; i++){    
            koleksi[i] = other.koleksi[i];
        }
    }

    public void printAll(){
        for(int i = 0; i < nEff; i++){
            System.out.println("Kendaraan " + (i+1) + ": ");
            koleksi[i].printInfo();
        }
    }

    /* Operator << */
    public void insert(Kendaraan K){
        if(nEff < size){
            koleksi[nEff] = K;
            nEff++;
        } else{
            System.out.println("Koleksi Penuh");
        }
    }

    /* Operator << */
    public void insert(KoleksiKendaraan other){
        for(int i = 0; i < other.nEff; i++){
            if(nEff < size){
                koleksi[nEff] = other.koleksi[i];
                nEff++;
            } else{
                System.out.println("Koleksi sudah Penuh");
                break;
            }
        }
    }
}
